File: ValidParentheses.py
class Solution:
    # @return a boolean
    def isValid(self, s):
        # Edge case fisrt !!
        left = ['(', '[', '{']
        right = [')', ']', '}']
        stack = []
        for i in s:
            if i in left:
                stack.append(i)
            if i in right:
                if stack and right.index(i) == left.index(stack[-1]):
                    stack.pop()
                else:
                    return False
        if not stack:
            return True
        else:
            return False


    # @return a boolean
    def isValid2(self, s):
        if len(s) <= 1:
            return False
        left, right = "([{", ")]}"
        stack = []
        for i in s:
            if i in left:
                stack.append(i)
            if i in right:
                if stack and right.index(i) == left.index(stack[-1]):
                    stack.pop()
                else:
                    return False
        return len(stack) == 0
                
        
if __name__ == '__main__':
    so = Solution()
    s = "["
    print so.isValid(s)

File: CPP/Second/valid_parentheses.cc
// https://oj.leetcode.com/problems/valid-parentheses/

// Given a string containing just the characters '(', ')', '{', '}', '[' and ']',
// The brackets must close in the correct order, "()" and "()[]{}" are all valid

#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        string left = "([{";
        string right = ")]}";
        stack<char> st;
        for (auto c: s) {
            if (left.find(c) != string::npos) {
                st.push(c);
            } else {
                if (st.empty() || st.top() != left[right.find(c)])
                    return false;
                else
                    st.pop();
            }
        }
        return st.empty();
    }
};

int main(int argc, char *argv[])
{

    return 0;
}
